### ISSUE TITLE
Move ExceptionMappers out of sgv2-quarkus-common

### ISSUE THREAD

**Issue by li-boxuan** [stargate/stargate#2248]
`io.stargate.sgv2.api.common.exception` within `sgv2-quarkus-common` package defines a few exception mappers which are meant to be shared across different APIs. I am trying to implement a custom RuntimeExceptionMapper (see https://github.com/li-boxuan/cassandra-dynamoDB-adapter/pull/3) but unfortunately, Quarkus **sometimes** picks up the  RuntimeExceptionMapper from `sgv2-quarkus-common` package in a non-deterministic way.

I tried the priority annotation as suggested by https://github.com/quarkusio/quarkus/issues/7883 but that didn't work. The only way I can think of is to move ExceptionMappers out of sgv2-quarkus-common - maybe another package, or copy the mappers to individual API packages?

**Comment by ivansenic** [comment]
You can exclude it, see https://quarkus.io/guides/cdi-reference#how-to-exclude-types-and-dependencies-from-discovery

**Comment by ivansenic** [comment]
@li-boxuan Let' me know if this works and if I can close the issue.

**Comment by li-boxuan** [comment]
@ivansenic Thanks! Yeah, I also tried that earlier but that didn't work. The error log seemed to suggest that the `RuntimeExceptionMapper` bean was excluded, but the `ServerExceptionMapper` was still registered. Let me try again and copy the error log.

**Comment by li-boxuan** [comment]
Excluding the bean doesn't work:

```yaml
quarkus:
  arc:
    exclude-types: io.stargate.sgv2.api.common.exception.*
```

The error log is:

```
ERROR [executor-thread-0] 2022-11-21 10:49:58,127 ResteasyReactiveRequestContext.java:629 - Request failed: java.lang.IllegalArgumentException: Unable to create class 'io.stargate.sgv2.api.common.exception.RuntimeExceptionMapper$GeneratedExceptionHandlerFor$RuntimeException$OfMethod$runtimeException'. To fix the problem, make sure this class is a CDI bean.
	at io.quarkus.resteasy.reactive.common.runtime.ArcBeanFactory.createInstance(ArcBeanFactory.java:41)
	at org.jboss.resteasy.reactive.server.core.RuntimeExceptionMapper.doGetExceptionMapper(RuntimeExceptionMapper.java:235)
	at org.jboss.resteasy.reactive.server.core.RuntimeExceptionMapper.getExceptionMapper(RuntimeExceptionMapper.java:214)
	at org.jboss.resteasy.reactive.server.core.RuntimeExceptionMapper.mapException(RuntimeExceptionMapper.java:76)
	at org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext.mapExceptionIfPresent(ResteasyReactiveRequestContext.java:340)
	at org.jboss.resteasy.reactive.server.handlers.ExceptionHandler.handle(ExceptionHandler.java:15)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:117)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:140)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:555)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: java.lang.NoSuchMethodException: io.stargate.sgv2.api.common.exception.RuntimeExceptionMapper$GeneratedExceptionHandlerFor$RuntimeException$OfMethod$runtimeException.<init>()
	at io.quarkus.arc.runtime.BeanContainerImpl$DefaultInstanceFactory.create(BeanContainerImpl.java:76)
	at io.quarkus.resteasy.reactive.common.runtime.ArcBeanFactory.createInstance(ArcBeanFactory.java:27)
	... 14 more
Caused by: java.lang.NoSuchMethodException: io.stargate.sgv2.api.common.exception.RuntimeExceptionMapper$GeneratedExceptionHandlerFor$RuntimeException$OfMethod$runtimeException.<init>()
	at java.base/java.lang.Class.getConstructor0(Class.java:3585)
	at java.base/java.lang.Class.getDeclaredConstructor(Class.java:2754)
	at io.quarkus.arc.runtime.BeanContainerImpl$DefaultInstanceFactory.create(BeanContainerImpl.java:68)
	... 15 more
```

Even worse, this error appears non-deterministically. Whenver my own `RuntimeExceptionMapper` gets picked up, everything works fine. Whenever the `RuntimeExceptionMapper` within `sgv2-quarkus-common` gets picked up, I saw the above error.

**Comment by ivansenic** [comment]
Let' me try to reproduce this. Thanks for the heads up.

**Comment by tatu-at-datastax** [comment]
Aside from excluding auto-registered mappers, perhaps there is a way for us to require explicit registration by APIs. That is, disable auto-registration by default, require APIs to explicitly add mapper(s) they want.

**Comment by ivansenic** [comment]
OK can reproduce, filed as bug on Quarkus project: https://github.com/quarkusio/quarkus/issues/29403

**Comment by ivansenic** [comment]
@li-boxuan The Quarkus can not make those optional.. I will try to move them to the JAX-RS style and add a configuration option to disable them.. I guess this should work.. Is this something that is blocking you at the moment?

**Comment by li-boxuan** [comment]
@ivansenic Sounds good! No, it’s not blocking me

**Comment by ivansenic** [comment]
Fix coming in https://github.com/quarkusio/quarkus/pull/29459, will wait for this to be released and make mappers optional..

**Comment by ivansenic** [comment]
Should be fixed in next release.

**Comment by ivansenic** [comment]
@li-boxuan Just specify `stargate.exception-mappers.enabled=false` prop in the `application.yaml`.. If you need to send a custom response on the missing token, let me know and I can tell you how to implement this. You basically need to register a `ChallengeSender` implementation in your app..

**Comment by li-boxuan** [comment]
@ivansenic I tried out 2.0.4 but unfortunately, it somehow doesn't work for me: https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24

**Comment by ivansenic** [comment]
> @ivansenic I tried out 2.0.4 but unfortunately, it somehow doesn't work for me: [li-boxuan/dynamoDB-adapter@8ef222c](https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24)

@li-boxuan Wrote a response on the PR: https://github.com/stargate/dynamoDB-adapter/pull/3#issuecomment-1365034746

**Comment by li-boxuan** [comment]
@ivansenic Thanks for taking a look! The PR you looked at was out-of-date. I did have `stargate.exception-mappers.enabled=false` here: https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24#diff-b4c0421bcb3280a8e738550c830dba197057c515bcdd9579a52cbfe9adca2061R15

**Comment by ivansenic** [comment]
> @ivansenic Thanks for taking a look! The PR you looked at was out-of-date. I did have `stargate.exception-mappers.enabled=false` here: [li-boxuan/dynamoDB-adapter@8ef222c#diff-b4c0421bcb3280a8e738550c830dba197057c515bcdd9579a52cbfe9adca2061R15](https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24#diff-b4c0421bcb3280a8e738550c830dba197057c515bcdd9579a52cbfe9adca2061R15)

And it's still not working?

**Comment by li-boxuan** [comment]
>  it's still not working?

Unfortunately, no.

**Comment by ivansenic** [comment]
@li-boxuan But what exactly is not working, do you still get response back as `ApiError` model?

**Comment by li-boxuan** [comment]
> But what exactly is not working

I sometimes still observe the default `RuntimeExceptionMapper` rather than my own captures the exception. In my custom RuntimeExceptionMapper, [this](https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24#diff-2ef59b4cbe114687dc20dbedbfc2bbc9eb720838e9f3aa9e3a71026e5e03b091R21) includes string literal "ErrorId" in the error message. Then I assert it [here](https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24#diff-60d008b182308c48292bb5f2b14dc4bb3c2ad67005adcfbb17ff57096f96fe2bR53). The assertion sometimes passes but sometimes not. Whenever the assertion fails, I also see the log printed out by the default `RuntimeExceptionMapper`, which confirms that the default `RuntimeExceptionMapper` is not disabled.

**Comment by ivansenic** [comment]
> > But what exactly is not working
> 
> I sometimes still observe the default `RuntimeExceptionMapper` rather than my own captures the exception. In my custom RuntimeExceptionMapper, [this](https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24#diff-2ef59b4cbe114687dc20dbedbfc2bbc9eb720838e9f3aa9e3a71026e5e03b091R21) includes string literal "ErrorId" in the error message. Then I assert it [here](https://github.com/li-boxuan/dynamoDB-adapter/commit/8ef222c787356e4b8fb86eddce86d9dc13150c24#diff-60d008b182308c48292bb5f2b14dc4bb3c2ad67005adcfbb17ff57096f96fe2bR53). The assertion sometimes passes but sometimes not. Whenever the assertion fails, I also see the log printed out by the default `RuntimeExceptionMapper`, which confirms that the default `RuntimeExceptionMapper` is not disabled.

Something seems off with the `testDescribeTable` how are you catching an exception by calling a `proxyClient`, as a result of the API invocation you should get the JSON back? How do you transform this to an exception? Can you provide a stack trace that proves that the common RuntimeExceptionMapper is used? 

Thing is that I have tests inside stargate/stargate that confirms thet are shutdown, and I plus have tests in antoher projects that also disables them and it's all fine. I though it maybe something specific for `RuntimeException`, but I test exactly that one here: https://github.com/stargate/stargate/blob/main/apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/exception/DisabledExceptionMappersTest.java

**Comment by li-boxuan** [comment]
> Something seems off with the testDescribeTable how are you catching an exception by calling a proxyClient, as a result of the API invocation you should get the JSON back? How do you transform this to an exception?

That's a long story :) In short, the client I used in the integration tests can only understand error responses in a certain format. It only understands my custom RuntimeExceptionMapper but not the default one in the quarkus-common library. My test verifies this behavior but I understand it is insufficient to convince people other than myself.

> I plus have tests in another projects that also disables them and it's all fine

Is that a public project I can try out as well? The chance is low but maybe the test could fail after more repetitive runs.

Btw do you see any problem of my `application.yaml`? Maybe I did something stupid in the config?

Thank you for your patience and I wish you a happy new year.

**Comment by ivansenic** [comment]
> Is that a public project I can try out as well? The chance is low but maybe the test could fail after more repetitive runs.
Unfortunately no.

> Btw do you see any problem of my `application.yaml`? Maybe I did something stupid in the config?
I can not see any issue there.

> Thank you for your patience and I wish you a happy new year.
Same for you :)
